### PR TITLE
bcc-python: support attach_func() and detach_func()

### PR DESCRIPTION
**Pull request by chenyuezhou**
Add `attach_func()` and `detach_func()`.

Add a usage sample.(`sockmap.py`)

### CODE REVIEW

**Comment by yonghong-song** [comment]
[buildbot, test this please]

File: docs/reference_guide.md [hunk 1]
```diff
@@ -1847,6 +1849,42 @@ Examples in situ:
 [search /examples](https://github.com/iovisor/bcc/search?q=attach_xdp+path%3Aexamples+language%3Apython&type=Code),
 [search /tools](https://github.com/iovisor/bcc/search?q=attach_xdp+path%3Atools+language%3Apython&type=Code)
 
+### 10. attach_func()
+
+Syntax: ```BPF.attach_func(fn, attach_type [, attachable_fd [, flags)```
```
Comment (yonghong-song) [inline docs/reference_guide.md:1854]: I think we can have a signature like
```
BPF.attach_func(fn, attachable_fd, attach_type [, flags])
```
Basically make attachable_fd mandatory and flags optional for python APIs.
The only case that an attchable_fd is not needed (must be 0) is for net_namespace. I think it is okay to require it and this makes argument sequence consistent with C++ APIs.

File: docs/reference_guide.md [hunk 2]
```diff
@@ -1847,6 +1849,42 @@ Examples in situ:
...
+
+Attaches a BPF function of the specified type.
```
Comment (yonghong-song) [inline docs/reference_guide.md:1856]: Attaches a BPF function of the specified type to a particular attachable_fd. if the attach_type is BPF_FLOW_DISSECTOR, the function is expected to attach to `current` net namespace and `attachable_fd` must be 0.

File: docs/reference_guide.md [hunk 3]
```diff
@@ -1847,6 +1849,42 @@ Examples in situ:
...
+
+Note. When attached to "global" hooks (xdp, tc, lwt, cgroup). If the "BPF function" is no longer needed after the program terminates, be sure to call `detach_func` when the program exits.
+
+Examples in situ:
+
+[search /examples](https://github.com/iovisor/bcc/search?q=attach_func+path%3Aexamples+language%3Apython&type=Code),
+
+### 11. detach_func()
+
+Syntax: ```BPF.detach_func(fn, detach_type [, target_fd)```
```
Comment (yonghong-song) [inline docs/reference_guide.md:1873]: Again, let us do ```BPF.detach_func(fn, attachable_fd, attach_type)```, similar to C++ APIs.

File: examples/networking/sockmap.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    return ret;
+}
+'''
+args = parser.parse_args()
+bpf = BPF(text=bpf_text)
+func_sock_ops = bpf.load_func("bpf_sockhash", bpf.SOCK_OPS)
+func_sock_redir = bpf.load_func("bpf_redir", bpf.SK_MSG)
+# raise if error
+fd = os.open(args.cgroup, os.O_RDONLY)
+map_fd = lib.bpf_table_fd(bpf.module, "sock_hash")
```
Comment (yonghong-song) [inline examples/networking/sockmap.py:115]: I got a failure like below with python3.6,
```
$ sudo python3.6 sockmap.py -c /sys/fs/cgroup/system.slice/tmp.service/
Traceback (most recent call last):
  File "sockmap.py", line 115, in <module>
    map_fd = lib.bpf_table_fd(bpf.module, "sock_hash")
ctypes.ArgumentError: argument 2: <class 'TypeError'>: wrong type
```

Changing "sock_hash" to b"sock_hash" fixed the issue. Probably some python byte array thing.

File: src/python/bcc/__init__.py [hunk 5]
```diff
@@ -544,6 +581,26 @@ def __delitem__(self, key):
     def __iter__(self):
         return self.tables.__iter__()
 
+    @staticmethod
+    def attach_func(fn, attach_type, attachable_fd=0, flags=0):
+        if not isinstance(fn, BPF.Function):
+            raise Exception("arg 1 must be of type BPF.Function")
+
+        res = lib.bpf_prog_attach(fn.fd, attachable_fd, attach_type, flags)
```
Comment (yonghong-song) [inline src/python/bcc/__init__.py:589]: Probably we should add bpf_prog_attach/detach2 to libbcc.py to make it explicit?

Comment (chenyuezhou) [reply src/python/bcc/__init__.py:589]: added.

File: examples/networking/sockmap.py (hunk 4 above)
Comment (chenyuezhou) [reply examples/networking/sockmap.py]: thanks, fixed.

File: docs/reference_guide.md (hunk 1 above)
Comment (chenyuezhou) [reply docs/reference_guide.md]: adjusted.

File: docs/reference_guide.md (hunk 2 above)
Comment (chenyuezhou) [reply docs/reference_guide.md]: thanks, adjusted.

File: docs/reference_guide.md (hunk 3 above)
Comment (chenyuezhou) [reply docs/reference_guide.md]: adjusted.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

File: examples/networking/sockmap.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+import sys
+import time
+import atexit
+import argparse
+
+from bcc import BPF, lib
+
+
+examples = """examples:
+    ./sockmap.py /root/cgroup # attach to /root/cgroup
```
Comment (yonghong-song) [inline examples/networking/sockmap.py:17]: ./sockmap.py -c /root/cgroup ?

File: examples/networking/sockmap.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+parser = argparse.ArgumentParser(
+        description="pipe data across multiple sockets",
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        epilog=examples)
+parser.add_argument("-c", "--cgroup", required=True,
+        help="Specify the cgroup address. Note. must be cgroup2")
+
+bpf_text = '''
+#include <net/sock.h>
+#include <bcc/proto.h>
```
Comment (yonghong-song) [inline examples/networking/sockmap.py:28]: `#include <bcc/proto.h>` is not needed.

**Review by yonghong-song — COMMENTED** [review]
LGTM. Just a few minor comments. Thanks!

File: examples/networking/sockmap.py (hunk 7 above)
Comment (chenyuezhou) [reply examples/networking/sockmap.py]: removed.

File: examples/networking/sockmap.py (hunk 6 above)
Comment (chenyuezhou) [reply examples/networking/sockmap.py]: adjusted.

File: docs/reference_guide.md [hunk 8]
```diff
@@ -1847,6 +1849,42 @@ Examples in situ:
 [search /examples](https://github.com/iovisor/bcc/search?q=attach_xdp+path%3Aexamples+language%3Apython&type=Code),
 [search /tools](https://github.com/iovisor/bcc/search?q=attach_xdp+path%3Atools+language%3Apython&type=Code)
 
+### 10. attach_func()
+
+Syntax: ```BPF.attach_func(fn, attachable_fd, attach_type [, flags)```
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply docs/reference_guide.md]: adjusted.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Pull request merged**